### DISCUSSION TITLE
VBAN stream from PC and recieve on ESP

### DISCUSSION THREAD

**Discussion by trackme518** [pschatzmann/arduino-audio-tools#1365] [Q&A] [answered]
Hi, I am trying to receive VBAN stream from PC on ESP32. The code compiles but I am not getting any audio from ESP (output to i2s). I have tested the generator -> i2s which is working. All the VBAN examples are for streaming audio from ESP to PC not the other way round. 

```
#include "AudioTools.h"
#include "AudioLibs/VBANStream.h"

AudioInfo info(44100, 2, 16);
VBANStream in;
I2SStream out;
StreamCopy copier(out, in);  // copies sound into i2s
```

```
 if (!out.begin(cfg)) stop();

  auto cfg_in = in.defaultConfig(RX_MODE);
  cfg_in.ssid = "mywifiname";
  cfg_in.password = "mywifipass";
  cfg_in.stream_name = "Stream1";
  //cfg_in.throttle_active = true;
  //cfg_in.throttle_correction_us = 0;  // optimize overload and underrun
  cfg_in.copyFrom(info);
  in.begin(cfg_in);

  Serial.println("VBAN WiFi receptor");
  Serial.print("Connected! IP address: ");
  Serial.println(WiFi.localIP());
```
```
void loop() {
  copier.copy();
}
```
I do get the message with my IP address so in the setup it seems the stream in is working - I can even do the VBAN ping but I still don't get any sound out...

Any help appreciated.

**Comment by pschatzmann** [comment]
Unfortunately I don't have any way to test this, so I suggest that you try to debug it yourself to find out what's going wrong.
Have a look what's happening in the receive_udp method in the VBANStream class:
- is it called at all ?
- does it fill the buffer ?

**Comment by pschatzmann** [comment]
I think in the beginRx method the connect is missing. Can you check if the following is working ?
```C++
    bool begin_rx(){
        start_wifi();
        if (WiFi.status() != WL_CONNECTED){
            LOGE("Wifi not connected");
            return false;
        }
        WiFi.setSleep(false); 
        // connect to target
        if (!udp.connect(cfg.target_ip, cfg.udp_port)){
            LOGE("Could not connect to target");
        }
        // handle data
        udp.onPacket([this](AsyncUDPPacket packet) {
             receive_udp(packet);
        });

        return true;
    }
```

**Reply by trackme518** [reply]
I have added the check inside VBANStream.h file inside the library and reflashed the sketch but I still do  not hear the sound output.

**Reply by pschatzmann** [reply]
Did you define the target adress and the port ?
You will need to check if receive_udp is executed. 
Maybe you add some distplay statments to figure out what's doing on...

**Reply by trackme518** [reply]
Hi, thank a lot for the help! I have defined the port  as `cfg_in.udp_port = 6980;` not the IP - I am receiving the VBAN stream and outputting into i2s, I think that I even can not define the target IP for incoming stream...I am defining the IP inside the voicemeeter app - which I have tested with mobile app VBAN receptor to verify that I am indeed sending the audio.  Do I need to rebuild the library somehow or just editing the .h file is enough?

**Reply by trackme518** [reply]
```
   bool begin_rx(){
    	LOGE("Logging into console");
        start_wifi();
        if (WiFi.status() != WL_CONNECTED){
            LOGE("Wifi not connected");
            return false;
        }
        WiFi.setSleep(false); 
        // connect to target
        if (!udp.connect(cfg.target_ip, cfg.udp_port)){
            LOGE("Could not connect to target");
        }
        // handle data
		udp.onPacket([this](AsyncUDPPacket packet) {
             LOGE("packet recieved");
			 receive_udp(packet);
        });

        return true;
    }
```
I am not seeing any print statements on udp recieve so I assume that the function is not even triggered.

**Reply by trackme518** [reply]
![save_voicemeeter](https://github.com/pschatzmann/arduino-audio-tools/assets/50262885/dc872f18-7b65-4a15-9e9c-cd0b5c397f63)
voicemeeter config

**Reply by trackme518** [reply]
I can ping the 192.168.1.111 and I get the reply in 6-11ms (expected - the device is indeed on the local network)

**Reply by trackme518** [reply]
Whole sketch:
```

#include "AudioTools.h"
#include "AudioLibs/VBANStream.h"

AudioInfo info(44100, 2, 16);
VBANStream in;
I2SStream out;
StreamCopy copier(out, in);

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  Serial.println("configuring output");

  auto cfg = out.defaultConfig(TX_MODE); 
  cfg.copyFrom(info);
  cfg.pin_bck = 35;   //BCK - default 14
  cfg.pin_ws = 37;    //LCK - default 15
  cfg.pin_data = 39;  //DIN - default 22
  cfg.sample_rate = 44100;
  out.begin(cfg);  //starts i2s

  Serial.println("configuring input");

  auto cfg_in = in.defaultConfig(RX_MODE);
  cfg_in.copyFrom(info);
  cfg_in.ssid = "WIFINAME";
  cfg_in.password = "WIFIPASS";
  cfg_in.stream_name = "Stream1";
  cfg_in.udp_port = 6980;

  in.begin(cfg_in);
  Serial.println("VBAN WiFi receptor");
  Serial.print("Connected! IP address: ");
  Serial.println(WiFi.localIP());
}

void loop() {
  copier.copy();
}
```

**Comment by pschatzmann** [comment]
Just looked at the ESP32 UDP example. Maybe the following might work better:

```

   bool begin_rx(){
    	LOGE("Logging into console");
        start_wifi();
        if (WiFi.status() != WL_CONNECTED){
            LOGE("Wifi not connected");
            return false;
        }
        WiFi.setSleep(false); 
        // connect to target
        if (!udp.listen(cfg.udp_port)){
            LOGE("Could not connect to target");
        }
        // handle data
		udp.onPacket([this](AsyncUDPPacket packet) {
             LOGE("packet recieved");
			 receive_udp(packet);
        });

        return true;
    }

```

**Reply by trackme518** [reply]
Ok, now I can hear extremely loud high pitch sound that stop and starts when I stop or start the stream + can I hear difference when I lower the sampling frequency. However, it is not the sound I am sending so probably some format problem.  It does not react on changing volume.

**Reply by pschatzmann** [reply]
The only supported format is PCM

**Reply by trackme518** [reply]
I think that you can not send anything else with VBAN anyway...from the screenshot in voicemeeter it shows 16bit PCM (when I switch to 24bit it is silent). I can also see that 44100Hz frequency is not listed in VBAN definition enum (`enum VBanSampleRates`) so I switched to 48000Hz. Now I am testing with `AudioInfo info(48000, 2, 24);` and that remains silent.

**Comment by pschatzmann** [comment]
No idea what this could be. Maybe VBAN is sending in network format (=big endian) and I am expecting little endian

**Reply by trackme518** [reply]
exactly my thoughts

**Reply by trackme518** [reply]
From the VBAN docs: ```
REM: VBAN protocol has been developed on x86 architecture. Data storage is consequently
following little-endian rules (least significant bytes are stored on first memory addresses).
```
source: https://vb-audio.com/Voicemeeter/VBANProtocol_Specifications.pdf

**Reply by trackme518** [reply]
I am getting this log: `[I] StreamCopy.h : 143 - StreamCopy::copy  540 -> 540 -> 540 bytes - in 1 hops`

**Reply by trackme518** [reply]
You have a note in the source code that reads `And set samplesPerPacket to 64 for 11025, 128 for 22050, or 256 for all else.` However, the "samplesPerPacket" is not a parameter inside the VBANStream class. Perhaps it means setting `StreamCopy copier(out, in, 256); ` I assume that the last parameter is for buffer size? This yeilds following:

```
[I] StreamCopy.h : 143 - StreamCopy::copy  256 -> 256 -> 256 bytes - in 1 hops
[I] StreamCopy.h : 143 - StreamCopy::copy  256 -> 256 -> 256 bytes - in 1 hops
[I] StreamCopy.h : 143 - StreamCopy::copy  28 -> 28 -> 28 bytes - in 1 hops
[I] StreamCopy.h : 143 - StreamCopy::copy  256 -> 256 -> 256 bytes - in 1 hops
[I] StreamCopy.h : 143 - StreamCopy::copy  256 -> 256 -> 256 bytes - in 1 hops
```
But it is still just high pitched artifacts noise. I tried to set the channels to 1 as well but it remains more or less the same.

**Reply by pschatzmann** [reply]
Maybe it would make sense to have the buffer filled to 50%  before switching it to active = available() reporting the available bytes

**Reply by trackme518** [reply]
To be clear I am using GY-PCM5102 I2S audio modul for i2s output (tested it's working).

**Reply by trackme518** [reply]
Also some time back I had a compilation error that I fixed by modifying this function:
```

        // limit output speed
        if (cfg.throttle_active){
        	//this method works
            throttle.delayFrames(samples / cfg.channels);
			//depraceted? dunno - but it does not work for esp32 s2 
			//throttle.delaySamples(samples / cfg.channels);
        }
```
I don't think it is related but noting it here for reference.

**Reply by pschatzmann** [reply]
I changed the logic a bit:
- Increased the buffer size to 1045
- available() is reporting a value only after the buffer has been filled to 50%
- issue error messages on buffer overflow

But I don't think that this will make a big difference...

**Comment by trackme518** [comment]
I am not sure where to look for this function.  There is:
```
    size_t readBytes(uint8_t* data, size_t byteCount) override {
        return rx_buffer.readArray(data, byteCount);
    }

    int available() {
        return rx_buffer.available();
    }
```
From where is sourced the `.available();`?

> Maybe it would make sense to have the buffer filled to 50% before switching it to active = available() reporting the available bytes

**Reply by pschatzmann** [reply]
just look at my latest commit

**Reply by trackme518** [reply]
I have copied the latest VBANStream.h file and replaced it in my local lib and reflashed the sketch. Unfortunately, It does not seem to change the behavior.

**Reply by pschatzmann** [reply]
but it should show copy 1024 now....

**Reply by trackme518** [reply]
Interestingly no - it shows `[I] StreamCopy.h : 143 - StreamCopy::copy  540 -> 540 -> 540 bytes - in 1 hops` even when i set `StreamCopy copier(out, in, 1024);` and `AudioInfo info(48000, 2, 16);`

**Reply by trackme518** [reply]
the same is true when i left the buffer size param empty like so `StreamCopy copier(out, in);`

**Comment by trackme518** [comment]
Even when I modify your line to: 
`if (bytes_received >= cfg.rx_buffer_count * DEFAULT_BUFFER_SIZE*2)` insted of `*0.5` (you are using `int max_write_size = DEFAULT_BUFFER_SIZE * 2;` at the top)
I only get `StreamCopy::copy  540 -> 540 -> 540 bytes - in 1 hops`

So I assume that a) check is not working b) check is always triggered once and later ignored (you set the flag to true than never reset it back). It does not seems to be triggered anyway (I tried to put LOGE fce there to see if it triggered and did not get anything out of it). I still don't understand how that should affect the function but I am just reporting back my findings.

**Comment by pschatzmann** [comment]
The idea was to report available() data only after the buffer has been filled by 50%. Before it will report 0, so nothing is copied.
As soon as the buffer had once more then 50%, available() will report the available data in the current buffer line of the n buffer which is DEFAULT_BUFFER_SIZE, that should be 1024 bytes, if you did not change this and the copier will limit the size to this..

**Reply by pschatzmann** [reply, marked as answer]
In order to reproduce this I had to buy the Talky Android app!
The worst error was, that the data was read from the wrong address. It is working much better now...
I also changed the buffer to use a thread save implementation and extended the API where you can define an output object, so that no copy is needed.

**Reply by trackme518** [reply]
VBAN reciever is free in lite version...https://play.google.com/store/apps/details?id=vbaudio.vbanreceptorlite&hl=en&gl=US you can use voicemeeter to stream from PC (and recieve on another PC or on mobile with mentioned lite version android app)

**Reply by trackme518** [reply]
Ok, I will try to download latest source and flash again and test it against esp32s2 - however, even at the start I was able to receive the stream on the mobile from voicemeeter without problem, so I am not sure what you have use that Talky Android App for.

**Reply by pschatzmann** [reply]
Yes, I know and I was testing my sender sketches with this. 
But you wanted to receive vban Audio on the ESP32 and there is not free version for this.
Since I do not have access to any Windows machine, this was the only option.

**Comment by trackme518** [comment]
Ah I see, sorry to hear that. I plan to implement the VBAN stream in Processing Java so we can have build for any OS.

**Comment by trackme518** [comment]
Hi, I can confirm it is now working on esp32s2. Minor issue remains that it does not conform to changing sampleRate to anything elese than 44100Hz - if I go lower or higher I will get artifacts. I can see you have function in the source code for dynamically changing this from the VBAN header but somehow it is not working - the same is true when I try to change it in config when starting the stream (ie hardcode in the beginning). 

```
 // update sample rate
      if (cfg.sample_rate != vbanSampleRate || cfg.channels != vbchannels) {
        cfg.sample_rate = vbanSampleRate;
        cfg.channels = vbchannels;
        setAudioInfo(cfg);
      }
```

**Reply by pschatzmann** [reply]
Strange, I did test this and it was working. Can you retry ? and mybe find out when it is not working...
I changed the sample rates and the number of channels...

**Reply by pschatzmann** [reply]
I updated the example to demo how the output can be informed about format changes in vban

**Reply by trackme518** [reply]
ok, thanks a lot i will test it in the evening!

**Reply by trackme518** [reply]
Hi, I just tested it and I can confirm it works - changing sample Rate is applied on the MCU side successfully. Thanks again!
